From 87c00b7cc3974fe77b3d4c975033dcc495bc1940 Mon Sep 17 00:00:00 2001
From: Ricardo Salveti <ricardo@foundries.io>
Date: Thu, 1 Oct 2020 21:05:28 -0300
Subject: [PATCH] arm: zynqmp: add support to handoff bl32 parameters

Add support to handoff bl32 related parameters to ATF, to allow SPL
booting both ATF and OP-TEE via fitimage.

Changes by Michael Scott.

Upstream-Status: Pending

Signed-off-by: Michael Scott <mike@foundries.io>
Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
---
 arch/arm/mach-zynqmp/handoff.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-zynqmp/handoff.c b/arch/arm/mach-zynqmp/handoff.c
index 64e5320acc..c36290539e 100644
--- a/arch/arm/mach-zynqmp/handoff.c
+++ b/arch/arm/mach-zynqmp/handoff.c
@@ -71,6 +71,7 @@ struct bl31_params *bl2_plat_get_bl31_params(uintptr_t bl32_entry,
 					     uintptr_t fdt_addr)
 {
 	struct xfsbl_atf_handoff_params *atfhandoffparams;
+	uint8_t entr = 0;
 
 	atfhandoffparams = (void *)CONFIG_SPL_TEXT_BASE;
 	atfhandoffparams->magic[0] = 'X';
@@ -78,13 +79,21 @@ struct bl31_params *bl2_plat_get_bl31_params(uintptr_t bl32_entry,
 	atfhandoffparams->magic[2] = 'N';
 	atfhandoffparams->magic[3] = 'X';
 
-	atfhandoffparams->num_entries = 0;
+	if (bl32_entry) {
+		atfhandoffparams->partition[entr].entry_point = bl32_entry;
+		atfhandoffparams->partition[entr].flags = (FSBL_FLAGS_EL1 <<
+						       FSBL_FLAGS_EL_SHIFT) |
+						       (FSBL_FLAGS_SECURE <<
+						        FSBL_FLAGS_TZ_SHIFT);
+		entr++;
+	}
 	if (bl33_entry) {
-		atfhandoffparams->partition[0].entry_point = bl33_entry;
-		atfhandoffparams->partition[0].flags = FSBL_FLAGS_EL2 <<
+		atfhandoffparams->partition[entr].entry_point = bl33_entry;
+		atfhandoffparams->partition[entr].flags = FSBL_FLAGS_EL2 <<
 						       FSBL_FLAGS_EL_SHIFT;
-		atfhandoffparams->num_entries++;
+		entr++;
 	}
+	atfhandoffparams->num_entries = entr;
 
 	writel(CONFIG_SPL_TEXT_BASE, &pmu_base->gen_storage6);
 
-- 
2.28.0

